QueueProcessors/QueueProcessor/queue_processor.py

```diff
@@ -424,96 +427,31 @@ def retry_run(reduction_run, retry_in):
-        """ Set the listener class. """
-        self._listener = listener
-
-    def get_connection(self, listener=None):
-        """ Return the connection and connect if not already connected. """
-        if listener is None and not self._client_only:
-            if self._listener is None:
-                listener = Listener(self)
-                self._listener = listener
-            else:
-                listener = self._listener
-
-        logger.info("[%s] Connecting to %s", self._consumer_name, str(self._brokers))
-
-        connection = stomp.Connection(host_and_ports=self._brokers, use_ssl=self._use_ssl)
-        if not self._client_only:
-            connection.set_listener(self._consumer_name, listener)
-        connection.start()
-        connection.connect(self._user, self._password, wait=False)
-
-        time.sleep(0.5)
-        return connection
-
-    def connect(self):
-        """ Connect and subscribe to the queues. """
-        if self._connection is None or not self._connection.is_connected():
-            self._disconnect()
-            self._connection = self.get_connection()
-
-        for queue in self._topics:
-            logger.info("[%s] Subscribing to %s", self._consumer_name, queue)
-            self._connection.subscribe(destination=queue, id=1, ack='auto')
-
-    def _disconnect(self):
-        """ Disconnect form the server. """
-        if self._connection is not None and self._connection.is_connected():
-            self._connection.disconnect()
-        self._connection = None
-        logger.info("[%s] Disconnected", self._consumer_name)
-
-    def stop(self):
-        """ Stop the connection. """
-        self._disconnect()
-        if self._connection is not None:
-            self._connection.stop()
-        self._connection = None
-
-    def send(self, destination, message, persistent='true', priority='4', delay=None):
-        """ Send a message to a queue. """
-        if self._connection is None or not self._connection.is_connected():
-            self._disconnect()
-            self._connection = self.get_connection()
-
-        headers = {}
-        if delay:
-            headers['AMQ_SCHEDULED_DELAY'] = str(delay)
-        self._connection.send(destination, message, persistent=persistent, priority=priority,
-                              headers=headers)
-        logger.debug("[%s] send message to %s", self._consumer_name, destination)
-
+def setup_connection(consumer_name):
+    """ Starts the ActiveMQ connection and registers the event listener """
+    logger.info("Starting autoreduce queue connection")
+    # Connect to ActiveMQ
+    activemq_client = QueueClient()
+    activemq_client.connect()
+
+    # Register the event listener
+    conn = activemq_client.get_connection()    
+    listener = Listener(activemq_client)
+    conn.set_listener(consumer_name, listener)
+    time.sleep(0.5)
+
+    # Subscribe to the queues
+    destinations = ['/queue/DataReady',
```

Comment:
We might want to move this to the `QueueClient`. I don't think it has much overhead and is probably required for 99% of cases in which we use `activemq`. Can you refactor the destination subscription to the client as well please and make sure it's called on initialisation.

Reply:
I'm not quite sure what you mean when you say "Can you refactor the destination subscription to the client as well please and make sure it's called on initialisation." Queue subscription is already directly invoked by the main method.

Reply:
So my thought behind this is that when we use `ActiveMQ`, (e.g. `QueueClient`) we will always want to subscribe to the various queues  (`'/queue/DataReady'`,`'/queue/ReductionStarted'`, etc). I can't think of a use case currently were we don't require queue subscription (expected some of the test cases.

As such we should move the functionality that is in this class to be in the QueueClient class instead. However, seeing as this change is probably going to affect all areas of the code base, on reflection this is probably out of scope for this issue. I'll create a new issue and assign you to it.

Reply:
See created issue #114